hub/util/merge.py

```diff
@@ -0,0 +1,321 @@
+):
+    conflict_indexes = []
+    for id in target_id_changes_commit_map:
+        target_commit_ids = target_id_changes_commit_map[id]
+        original_commit_ids = original_id_changes_commit_map[id]
+        set_original_commit_ids = set(original_commit_ids)
+        idx = None
+        for i, item in enumerate(target_commit_ids):
+            if item in set_original_commit_ids:
+                idx = i
+                break
+
+        # if no id is common or if a commit id other than the most recent commit_id is in common, there's a conflict
+        if idx is None or idx > 0:
+            target_idx = target_id_to_index_map[id]
+            original_idx = original_id_to_index_map[id]
+            conflict_indexes.append((original_idx, target_idx))
+    return conflict_indexes
+
+
+def process_tensor(
+    tensor_name: str,
+    dataset,
+    target_dataset,
+    original_node: CommitNode,
+    target_node: CommitNode,
+    lca_node: CommitNode,
+    new_samples_dict,
+    conflict_samples_dict,
+    conflict_resolution: Optional[str] = None,
+):
+    id_tensor_name = get_sample_id_tensor_key(tensor_name)
+    target_id_tensor = target_dataset[id_tensor_name]
+    original_id_tensor = dataset[id_tensor_name]
+
+    target_id_changes_commit_map = get_changes_commit_ids_for_node(
+        target_dataset, tensor_name, target_node, lca_node
+    )
+
+    original_id_changes_commit_map = get_changes_commit_ids_for_node(
+        dataset, tensor_name, original_node, lca_node
+    )
+
+    original_ids = original_id_tensor.numpy().flatten()
+    original_id_to_index_map = {id: idx for idx, id in enumerate(original_ids)}
+
+    target_ids = target_id_tensor.numpy().flatten()
+    target_id_to_index_map = {id: idx for idx, id in enumerate(target_ids)}
+
+    new_elements_ids = set(target_ids) - set(original_ids)
+
+    new_indexes = get_new_indexes(
+        new_elements_ids, target_id_changes_commit_map, target_id_to_index_map
+    )
+    new_samples_dict[tensor_name] = new_indexes
+
+    if conflict_resolution is None or conflict_resolution == "theirs":
+        conflict_indexes = find_conflicts(
+            original_id_changes_commit_map,
+            target_id_changes_commit_map,
+            original_id_to_index_map,
+            target_id_to_index_map,
+        )
+        if conflict_indexes:
+            conflict_samples_dict[tensor_name] = conflict_indexes
+
+
+def merge_tensor(
+    tensor_name: str, dataset, target_dataset, new_samples_dict, conflict_samples_dict
+):
+    original_tensor = dataset[tensor_name]
+    target_tensor = target_dataset[tensor_name]
+
+    new_indexes = new_samples_dict[tensor_name]
+    for index in new_indexes:
+        original_tensor.append(target_tensor[index])
+
+    if tensor_name in conflict_samples_dict:
+        conflict_indexes = conflict_samples_dict[tensor_name]
+        for original_idx, target_idx in conflict_indexes:
+            original_tensor[original_idx] = target_tensor[target_idx]
```

Comment:
oh you can reindex or is it actually doing copy of the element?

Reply:
it's a copy, the reason for this is to make streaming faster while loading data, otherwise a bunch of extra data will be fetched from chunks from which only 1 or few samples are reused.

Reply:
ok